data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/RemotePeerForwarder.java

```diff
@@ -75,6 +81,20 @@ public Collection<Record<Event>> forwardRecords(final Collection<Record<Event>>
         return recordsToProcessLocally;
     }
 
+    public Collection<Record<Event>> receiveRecords() {
+        // TODO: Make the read timeout configurable? This is the default read batch delay in PipelineConfiguration
+        final Map.Entry<Collection<Record<Event>>, CheckpointState> readResult =
+                peerForwarderReceiveBuffer.read(READ_BATCH_DELAY);
+
+        final Collection<Record<Event>> records = readResult.getKey();
+        final CheckpointState checkpointState = readResult.getValue();
+
+        // Checkpoint the current batch read from the buffer after reading from buffer
+        peerForwarderReceiveBuffer.checkpoint(checkpointState);
```

Comment:
We normally checkpoint the events after the entire processor chain ends. Maybe we should move it to the primary buffer and then checkpoint from the server buffer?

Reply:
I'm not sure I understood this. How is this related to primary buffer? Primary buffer creates a checkpoint once all the events are processed by processors and sinks. 

Let's say source writes 10 records to buffer and peer forwarder forwards 5 to peers and processes 5 locally. After these 5 events are processed by entire processor chain locally we checkpoint for entire 10 events we initially received.